sdk/python/pvsite_datamodel/write/forecast.py

```diff
@@ -6,37 +6,62 @@
 
 import numpy.typing as npt
 import pandas as pd
+import pandera as pa
 import sqlalchemy.orm as sa_orm
 
 from pvsite_datamodel.read.site import get_site_by_uuid
 from pvsite_datamodel.sqlmodels import ForecastSQL, ForecastValueSQL
 from pvsite_datamodel.write.datetime_intervals import get_or_else_create_datetime_interval
-from pvsite_datamodel.write.utils import WrittenRow
+from pvsite_datamodel.write.utils import UUIDV4, WrittenRow
+
+
+class ForecastValuesSchema(pa.SchemaModel):
+    """Schema for the dataframe used by the insert_forecast_values function."""
+
+    target_start_utc: pa.typing.Series[pd.DatetimeTZDtype] = pa.Field(
+        dtype_kwargs={"unit": "ns", "tz": "UTC"},
+        description="The start of the target time period for the forecast value."
+    )
+
+    target_end_utc: pa.typing.Series[pd.DatetimeTZDtype] = pa.Field(
+        dtype_kwargs={"unit": "ns", "tz": "UTC"},
+        description="The end of the target time period for the forecast value."
+    )
+    forecast_kw: pa.typing.Series[pa.dtypes.Float] = pa.Field(
+        ge=0,
+        description="The forecast generation for the site in kilowatts."
+    )
+    site_uuid: pa.typing.Series[UUIDV4] = pa.Field(
+        description="The UUID of the site for which the forecast value was created."
+    )
 
 
 def insert_forecast_values(
-    session: sa_orm.Session, df_forecast_values: pd.DataFrame, forecast_version: str = "0.0.0"
+        session: sa_orm.Session, forecast_values_df: pa.typing.DataFrame[ForecastValuesSchema],
+        forecast_version: str = "0.0.0"
 ) -> list[WrittenRow]:
     """Insert a dataframe of forecast values into the database.
 
     :param session: sqlalchemy session for interacting with the database
-    :param df_forecast_values: pandas dataframe with columns
-        ["target_start_utc", "target_end_utc", "forecast_kw", "pv_uuid"]
+    :param forecast_values_df: dataframe conforming to ForecastValuesSchema
     :param forecast_version: the version of the model used to create the forecast
     :return list[WrittenRow]: list of added rows to DB
     """
+    # Validate the incoming dataframe
+    ForecastValuesSchema.validate(forecast_values_df)
+
     # Track rows added to DB
     written_rows: list[WrittenRow] = []
 
     # Loop over all the unique sites that have got forecast values
-    sites: npt.ndarray[uuid.UUID] = df_forecast_values["pv_uuid"].unique()
-
+    sites: npt.ndarray[uuid.UUID] = forecast_values_df["site_uuid"].unique()
     for site_uuid in sites:
+
         # Check whether the site id exits in the table, otherwise return an error
         get_site_by_uuid(session=session, site_uuid=site_uuid)
 
         # Create a forcast (sequence) for the site, and write it to db
-        forecast = ForecastSQL(
+        forecast: ForecastSQL = ForecastSQL(
```

Comment:
Why the "double" typing?

Reply:
Thats, probably left over from me, maybe it helps with PyCharm type hinting, but im not sure

Reply:
Ah I see that would make sense. There is probably a way to setup PyCharm so that it doesn't need it!